jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdAuthenticator.java

```diff
@@ -0,0 +1,481 @@
+                    {
+                        MultiMap<String> formParameters = new MultiMap<>();
+                        baseRequest.extractFormParameters(formParameters);
+                        session.setAttribute(__J_POST, formParameters);
+                    }
+                }
+            }
+
+            // send the the challenge
+            String challengeUri = getChallengeUri(request);
+            LOG.debug("challenge {}->{}", session.getId(), challengeUri);
+            int redirectCode = (baseRequest.getHttpVersion().getVersion() < HttpVersion.HTTP_1_1.getVersion() ? HttpServletResponse.SC_MOVED_TEMPORARILY : HttpServletResponse.SC_SEE_OTHER);
+            baseResponse.sendRedirect(redirectCode, response.encodeRedirectURL(challengeUri));
+
+            return Authentication.SEND_CONTINUE;
+        }
+        catch (IOException e)
+        {
+            throw new ServerAuthException(e);
+        }
+    }
+
+    public boolean isJSecurityCheck(String uri)
+    {
+        int jsc = uri.indexOf(__J_SECURITY_CHECK);
+
+        if (jsc < 0)
+            return false;
+        int e = jsc + __J_SECURITY_CHECK.length();
+        if (e == uri.length())
+            return true;
+        char c = uri.charAt(e);
+        return c == ';' || c == '#' || c == '/' || c == '?';
+    }
+
+    public boolean isErrorPage(String pathInContext)
+    {
+        return pathInContext != null && (pathInContext.equals(_errorPath));
+    }
+
+    private String getRedirectUri(HttpServletRequest request)
+    {
+        final StringBuffer redirectUri = new StringBuffer(128);
+        URIUtil.appendSchemeHostPort(redirectUri, request.getScheme(),
+            request.getServerName(), request.getServerPort());
+        redirectUri.append(request.getContextPath());
+        redirectUri.append(__J_SECURITY_CHECK);
+        return redirectUri.toString();
+    }
+
+    protected String getChallengeUri(HttpServletRequest request)
+    {
+        HttpSession session = request.getSession();
+        String antiForgeryToken;
+        synchronized (session)
+        {
+            antiForgeryToken = (session.getAttribute(__CSRF_TOKEN) == null)
+                ? new BigInteger(130, new SecureRandom()).toString(32)
+                : (String)session.getAttribute(__CSRF_TOKEN);
+            session.setAttribute(__CSRF_TOKEN, antiForgeryToken);
+        }
+
+        // any custom scopes requested from configuration
+        StringBuilder scopes = new StringBuilder();
+        for (String s : _configuration.getScopes())
+        {
+            scopes.append("%20" + s);
+        }
+
+        return _configuration.getAuthEndpoint() +
+            "?client_id=" + _configuration.getClientId() +
+            "&redirect_uri=" + getRedirectUri(request) +
+            "&scope=openid" + scopes +
+            "&state=" + antiForgeryToken +
+            "&response_type=code";
```

Comment:
I'm not sure this is right.
As this is concatenating query string parameters, the parameter values must be encoded.
You cannot have something like `http://authHost/?client_id=1&redirect_uri=http://walkaboutcreek/?a=a&b=b...`
Must be like: `http://authHost/?client_id=1&redirect_uri=http%3A%2F%2F...`.